crash when logging a unloaded PM only cache as BM from search

**Describe the bug:**
Using a non- premium account, I enter in GC# of a premium cache and receive "error" message. But the option to log cache. Clicking "log" shuts down the app. 

**To Reproduce:**
Steps to reproduce the behavior:
1. From basic account, Go to search and type in GC# of premium cache. Search. 
2. Dialog box appears for Premium Member Only. 
3. Click LOG VISIT

**Actual behavior/state after performing these steps:**
App Crashes. 

**Expected behavior/state after performing these steps:**
A dialog box should appear to allow a basic membership to log a Premium Cache. 

**Version of c:geo used:**
2020.10.29

**Is the problem reproducible:**
Yes

**Screenshots:**

![Screenshot_20201030-202521_cgeo](https://user-images.githubusercontent.com/71992117/97766885-f92dd080-1aee-11eb-822a-8829f78c5dab.jpg)


**System information:**
```
--- System information ---
Device: SM-S767VL (j7topeltetfnvzw, samsung)
Android version: 9
Android build: PPR1.180610.011.S767VLUDS8BTI2
c:geo version: 2020.10.29
Google Play services: enabled - 20.39.15 (100300-335085812)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide caches: own/found 
Hide waypoints: original 
HW acceleration: enabled (default state)
System language: en_US
System date format: M/d/yy
Debug mode active: no
System internal c:geo dir: /data/user/0/cgeo.geocaching (1.0 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (1.0 GB free) external non-removable
Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (1.0 GB free) external non-removable
Database: /data/user/0/cgeo.geocaching/databases/data (4.1 MB) on system internal storage
GPX import path: /storage/emulated/0/cgeo/gpx
GPX export path: /storage/emulated/0/cgeo/gpx
Offline maps path: /storage/emulated/0/cgeo/maps
Map render theme path: 
Live map mode: true
Global filter: display all caches
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
   geocaching.com: Logged in (Login OK) / BASIC
Geocaching.com date format: MMM/dd/yyyy
Installed c:geo plugins:  none
BRouter connection available: false
--- End of system information ---


**Additional context:**
e.g. Reference to other issues, projects, sources, etc.

Following up on a part of the log above:

10-31 21:09:26.630 3123 3123 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.List cgeo.geocaching.models.Geocache.getPossibleLogTypes()' on a null object reference
10-31 21:09:26.630 3123 3123 E AndroidRuntime: at cgeo.geocaching.log.LogCacheActivity.onCreate(LogCacheActivity.java:268)
10-31 21:09:26.630 3123 3123 E AndroidRuntime: at android.app.Activity.performCreate(Activity.java:7335)
10-31 21:09:26.630 3123 3123 E AndroidRuntime: at android.app.Activity.performCreate(Activity.java:7326)

Looking at the code, LogCacheActivity is written in a way that it needs a filled cache-object to work, and it expects this object to be in the CacheCache or the DB at time of being called:

https://github.com/cgeo/cgeo/blob/8d1eb842f2858125ea300c3d80661b1e725ce5c6/main/src/cgeo/geocaching/log/LogCacheActivity.java#L292-L295

In the current situation this object seems to be null. If my guess is right this problem can be reproduced when logged in as non-PM **only for caches which are not yet in the DB or local cache**.

I can trace back this behaviour using "git blame" back to the year 2013. E.g. here is a screenshot from a version in June2020. When was the last time this function was prooven to work fine?
![image](https://user-images.githubusercontent.com/6909759/97799277-5fb90880-1c2d-11eb-9e76-9bdc21a73a28.png)

I assume (?) that cgeo cannot load information for premium caches from the web. So in order to solve this one would need to make some assumptions on what should happen in log dialog in case we have no information about a geocache e.g. which log types shall be available. Also, access to GCConnector and LoggingManager all goes through Geocache object currently, so a redesign might be necessary.

This is all from a glance at the code. Anyone around knowing on how logging of non-PM-caches was implemented in the past and what the idea is here?

--- a/main/src/cgeo/geocaching/connector/gc/GCConstants.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCConstants.java
@@ -168,6 +168,13 @@ public final class GCConstants {
 
     static final String STRING_PREMIUMONLY_2 = "Sorry, the owner of this listing has made it viewable to Premium Members only.";
     static final String STRING_PREMIUMONLY_1 = "class=\"illustration lock-icon\"";
+    static final Pattern PATTERN_PREMIUMONLY_CACHETYPE = Pattern.compile("/app/ui-icons/sprites/cache-types\\.svg#icon-([^\"\\-]+)-?([^\"]+)?");
+    static final Pattern PATTERN_PREMIUMONLY_CACHENAME = Pattern.compile("<h1 class=\"heading-3\">(.+)</h1>");
+    static final Pattern PATTERN_PREMIUMONLY_GEOCODE = Pattern.compile("<li class=\"li__gccode\">([^<]+)");
+    static final Pattern PATTERN_PREMIUMONLY_DIFFICULTY = Pattern.compile("<span id=\"ctl00_ContentBody_lblDifficulty\"(?:.|\\s)*?<span>([0-5](?:[\\.,]5)?)</span>");
+    static final Pattern PATTERN_PREMIUMONLY_TERRAIN = Pattern.compile("<span id=\"ctl00_ContentBody_lblTerrain\"(?:.|\\s)*?<span>([0-5](?:[\\.,]5)?)</span>");
+    static final Pattern PATTERN_PREMIUMONLY_SIZE = Pattern.compile("<span id=\"ctl00_ContentBody_lblSize\"(?:.|\\s)*?<span>([^<]+)</span>");
+
     static final String STRING_UNPUBLISHED_OTHER = "you cannot view this cache listing until it has been published";
     static final String STRING_UNPUBLISHED_FROM_SEARCH = "class=\"UnpublishedCacheSearchWidget"; // do not include closing brace as the CSS can contain additional styles
     static final String STRING_UNKNOWN_ERROR = "An Error Has Occurred";

--- a/main/src/cgeo/geocaching/connector/gc/GCParser.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCParser.java
@@ -51,6 +51,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 import java.util.regex.Pattern;
 
@@ -386,7 +387,13 @@ public final class GCParser {
         final ImmutablePair<StatusCode, Geocache> parsed = parseCacheFromText(page, handler);
         // attention: parseCacheFromText already stores implicitly through searchResult.addCache
         if (parsed.left != StatusCode.NO_ERROR) {
-            return new SearchResult(parsed.left);
+            final SearchResult result = new SearchResult(parsed.left);
+
+            if (parsed.left == StatusCode.PREMIUM_ONLY) {
+                result.addAndPutInCache(Collections.singletonList(parsed.right));
+            }
+
+            return result;
         }
 
         final Geocache cache = parsed.right;
@@ -447,7 +454,21 @@ public final class GCParser {
         }
 
         if (pageIn.contains(GCConstants.STRING_PREMIUMONLY_1) || pageIn.contains(GCConstants.STRING_PREMIUMONLY_2)) {
-            return ImmutablePair.of(StatusCode.PREMIUM_ONLY, null);
+            final Geocache cache = new Geocache();
+            cache.setPremiumMembersOnly(true);
+            final MatcherWrapper matcher = new MatcherWrapper(GCConstants.PATTERN_PREMIUMONLY_CACHETYPE, pageIn);
+            if (matcher.find()) {
+                cache.setType(CacheType.getByWaypointType(matcher.group(1)));
+                if (Objects.equals(matcher.group(2), "disabled")) {
+                    cache.setDisabled(true);
+                }
+            }
+            cache.setName(TextUtils.getMatch(pageIn, GCConstants.PATTERN_PREMIUMONLY_CACHENAME, true, ""));
+            cache.setGeocode(TextUtils.getMatch(pageIn, GCConstants.PATTERN_PREMIUMONLY_GEOCODE, true, ""));
+            cache.setDifficulty(Float.parseFloat(TextUtils.getMatch(pageIn, GCConstants.PATTERN_PREMIUMONLY_DIFFICULTY, true, "0")));
+            cache.setTerrain(Float.parseFloat(TextUtils.getMatch(pageIn, GCConstants.PATTERN_PREMIUMONLY_TERRAIN, true, "0")));
+            cache.setSize(CacheSize.getById(TextUtils.getMatch(pageIn, GCConstants.PATTERN_PREMIUMONLY_SIZE, true, CacheSize.NOT_CHOSEN.id)));
+            return ImmutablePair.of(StatusCode.PREMIUM_ONLY, cache);
         }
 
         final String cacheName = TextUtils.stripHtml(TextUtils.getMatch(pageIn, GCConstants.PATTERN_NAME, true, ""));

--- a/tests/src-android/cgeo/geocaching/CgeoApplicationTest.java
+++ b/tests/src-android/cgeo/geocaching/CgeoApplicationTest.java
@@ -109,7 +109,7 @@ public class CgeoApplicationTest extends CGeoTestCase {
     public static Geocache testSearchByGeocode(final String geocode) {
         final SearchResult search = Geocache.searchByGeocode(geocode, null, true, null);
         assertThat(search).isNotNull();
-        if (Settings.isGCPremiumMember() || search.getError() == StatusCode.NO_ERROR) {
+        if (Settings.isGCPremiumMember() || search.getError() == StatusCode.NO_ERROR || search.getError() == StatusCode.PREMIUM_ONLY) {
             assertThat(search.getGeocodes()).containsExactly(geocode);
             return DataStore.loadCache(geocode, LoadFlags.LOAD_CACHE_OR_DB);
         }
@@ -159,20 +159,29 @@ public class CgeoApplicationTest extends CGeoTestCase {
             SearchResult search = Geocache.searchByGeocode(cache.getGeocode(), null, true, null);
             assertThat(search).isNotNull();
             assertThat(search.getGeocodes()).containsExactly(cache.getGeocode());
-            final Geocache searchedCache = search.getFirstCacheFromResult(LoadFlags.LOAD_CACHE_OR_DB);
+            Geocache searchedCache = search.getFirstCacheFromResult(LoadFlags.LOAD_CACHE_OR_DB);
             // coords must be null if the user is not logged in
             assertThat(searchedCache).isNotNull();
-            assert searchedCache != null; // eclipse bug
             assertThat(searchedCache.getCoords()).isNull();
 
-            // premium cache. Not visible to guests
+            // premium cache. Not fully visible to guests
             cache = new GC2JVEH();
 
             deleteCacheFromDBAndLogout(cache.getGeocode());
 
             search = Geocache.searchByGeocode(cache.getGeocode(), null, true, null);
             assertThat(search).isNotNull();
-            assertThat(search.getGeocodes()).isEmpty();
+            searchedCache = search.getFirstCacheFromResult(LoadFlags.LOAD_CACHE_OR_DB);
+            assertThat(searchedCache).isNotNull();
+            assertThat(searchedCache.getCoords()).isNull();
+            assertThat(searchedCache.getName()).isEqualTo(cache.getName());
+            assertThat(searchedCache.getDifficulty()).isEqualTo(cache.getDifficulty());
+            assertThat(searchedCache.getTerrain()).isEqualTo(cache.getTerrain());
+            assertThat(searchedCache.getGeocode()).isEqualTo(cache.getGeocode());
+            assertThat(searchedCache.getSize()).isEqualTo(cache.getSize());
+            assertThat(searchedCache.getType()).isEqualTo(cache.getType());
+            // it's not possible in guest sessions to distinguish whether a PM-only cache is disabled or archived
+            assertThat(searchedCache.isDisabled()).isEqualTo(cache.isDisabled() || cache.isArchived());
         });
     }
 

